evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -203,64 +235,104 @@ def transform(
 
         Args:
             X (pd.DataFrame, optional): Conditionally used to build datetime index.
-            y (pd.Series): Target variable to detrend and deseasonalize.
+            y (pd.Series or pd.DataFrame): Target variable to detrend and deseasonalize.
 
         Returns:
-            tuple of pd.DataFrame, pd.Series: The input features are returned without modification. The target
+            (Single series) pd.DataFrame, pd.Series: The list of input features are returned without modification. The target
+                variable y is detrended and deseasonalized.
+            (Multi series) pd.DataFrame, pd.DataFrame: The list of input features are returned without modification. The target
                 variable y is detrended and deseasonalized.
 
         Raises:
             ValueError: If target data doesn't have DatetimeIndex AND no Datetime features in features data
         """
         if y is None:
             return X, y
-        original_index = y.index
-        X, y = self._check_target(X, y)
-        self._check_oos_past(y)
 
-        y_in_sample = pd.Series([])
-        y_out_of_sample = pd.Series([])
-
-        # For partially and wholly in-sample data, retrieve stored results.
-        if self.trend.index[0] <= y.index[0] <= self.trend.index[-1]:
-            y_in_sample = self.residual[y.index[0] : y.index[-1]]
-
-        # For out of sample data....
-        if y.index[-1] > self.trend.index[-1]:
-            try:
-                # ...that is partially out of sample and partially in sample.
-                truncated_y = y[y.index.get_loc(self.trend.index[-1]) + 1 :]
-            except KeyError:
-                # ...that is entirely out of sample.
-                truncated_y = y
-
-            (
-                projected_trend,
-                projected_seasonality,
-            ) = self._project_trend_and_seasonality(truncated_y)
-
-            y_out_of_sample = infer_feature_types(
-                pd.Series(
-                    truncated_y - projected_trend - projected_seasonality,
-                    index=truncated_y.index,
-                ),
-            )
-        y_t = pd.concat([y_in_sample, y_out_of_sample])
-        y_t.index = original_index
-        return X, y_t
+        if isinstance(y, pd.Series):
+            y = y.to_frame()
+        detrending_list = []
+        # Iterate through each id group
+        for id in y.columns:
+            series_y = y[id]
+
+            if len(y.columns) > 1:
+                seasonality = self.seasonalities[id]
+                trend = self.trends[id]
+                residual = self.residuals[id]
+                period = self.periods[id]
+            else:
+                seasonality = list(self.seasonalities.values())[0]
```

Comment:
Had to add this here since I removed the class attributes and replaced them with dictionaries. As a result, in the pipeline i.e. for `test_pipeline_utils` there was a key error when transforming new data when we don't fit before transforming since the keys don't match up in the dictionary. I'm not sure if there's a better way to be doing this or if we need a way to be able to transform on different data in the multiseries case.

Reply:
It was failing in`evalml/tests/pipeline_tests/test_pipeline_utils.py::test_generate_pipeline_example`. I think from what I saw it was fitting and transforming, and then it was transforming again but was returning a key error since the id wasn't the same from the initial fitting

Reply:
Hmm - @jeremyliweishih, any thoughts? This feels like an edge case, but I can't follow the pipeline generation code well enough to understand why the target name isn't matching - it feels like this shouldn't actually be an issue.

Reply:
Not a blocking issue, but tracking it in #4295